Add `AccountRegistry` support for DAI `permit()` function

thomas-waite opened the pull request:
## Summary
This PR adds support to the `AccountRegistry.sol` for the DAI `permit()` function. 

The new DAI `permit()` function allows ERC20 spending approval to be granted via an offchain signature, rather than the sending of a transaction. 

Background on the `DAI.permit()` method: https://docs.makerdao.com/smart-contract-modules/dai-module/dai-detailed-documentation 

## Description
A new `AccountRegistry` behaviour contract, `Behaviour20200220.sol` has been added. This defines two methods with unique function signatures:

1. `deposit()` - convert DAI into zkDAI
2. `permit()`- call `DAI.permit()`

Given that `permit()` is currently not part of the ERC20 standard, and is a DAI specific method, it is importantly that the `AccountRegistry` supports both legacy approval flows and the new `permit()` approval flow. 

To this end, both the legacy `Behaviour20200106.deposit()` method and the new `Behaviour20200220.deposit()` method are available - they have different signatures as `Behaviour20200220.deposit()` takes two additional parameters `signature` and `nonce`. 

`signature` is an EIP712 signature that is passed to the `DAI.permit()` method. It is used to approve the `proxyContract` to spend DAI on behalf of the user, rather than having to send a `dai.approve()` transaction.  

`nonce` is a record of the number of `permit()` calls have been made and is used for signature replay protection in the DAI contract. 

### `Behaviour20200220.permit()`
This method has been given the visibility modifier of `public` so that users are able to directly call `permit()` and so revoke approval at a later date if they wish/need to.

thomas-waite committed:
feat(dev-utils): add EIP712 constants for sign permit fn

thomas-waite committed:
feat(aztec.js): add permit signing function
add associated test
tidy up other MetaMask comparison tests

thomas-waite committed:
build: update yarn.lock

thomas-waite committed:
feat(protocol): add DAI token contracts

thomas-waite committed:
refactor: make permit signing specific to DAI

thomas-waite committed:
feat(protocol): add Behaviour contract with permit functionality

thomas-waite committed:
test(protocol): add tests for permit behaviour contract

thomas-waite committed:
feat(protocol): add interface contract for ERC20 with permit

thomas-waite committed:
test(protocol): update nonce

thomas-waite committed:
feat(protocol): add new behvaiour migration script

thomas-waite committed:
feat(contract-artifacts): add permit behaviour to artifacts

thomas-waite committed:
feat(monorepo-scripts): add permit behaviour to addresses

joeandrews committed:
feat(sdk): update the sdk to detect permit from config

thomas-waite committed:
refactor(protocol): expose more params in permit(), so user can revoke
linting fixes

thomas-waite committed:
test(protocol): check that legacy deposit() still callable
add failure test

joeandrews committed:
feat(sdk): fix typo

ArnSch commented on packages/extension/src/ui/apis/asset/permitERC20Allowance.js:
```diff
@@ -0,0 +1,54 @@
+    } catch (e) {
+        error = e;
+    }
+    const exipry = Date.now() + 24 * 60 * 60;
```
```suggestion
    const expiry = Date.now() + 24 * 60 * 60;
```

(this is going to go anyways)

ArnSch commented on packages/protocol/contracts/AccountRegistry/epochs/20200220/Behaviour20200220.sol:
```diff
@@ -0,0 +1,146 @@
+    * @param signature - EIP712 signature supplied to permit() for granting approval
+    * @param nonce - permit nonce used for signature replay protection
+     */
+    function deposit(
```
As discussed, should make the permit call to transfer tokens to this contract first, then rely on `super.deposit()` to call existing deposit function in previous inherited behaviour to make changes explicit, and increase confidence that no changes in logic have occurred beyond expected ones.

thomas-waite replied:
Yeah good idea, have done

ArnSch reviewed (changes requested):
Looks good. We discussed changes to deposit function, and to make the permit timestamp much more constrained, but otherwise good to go.

joeandrews committed:
feat(sdk): sorking permit signature for dai

joeandrews committed:
feat(sdk): working permit

joeandrews committed:
feat(sdk): update interface for account registry

thomas-waite committed:
refactor(protocol): implement PR feedback
expose expiry as var in deposit() - to allow it to be made significantly more restrictive
add failure test for expired expiry
refactor new deposit() to call super.deposit()

thomas-waite committed:
refactor(protocol): update IAccountRegistryBehaviour interface contract
add missing param doc string

thomas-waite committed:
feat(contract-artifacts): add IERC20Permit artifact for cirlceci

Commit:
style(sdk): fix all non-feature lint errors

ArnSch commented:
@thomas-waite still one lint error (`expiry` declared but not used in `depositWithPermit`), but did not know whether that was WIP so left as is. Otherwise fixed lint issues.

joeandrews committed:
feat(sdk): add expiry to frontend

joeandrews commented:
:tada: This PR is included in version 0.17.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

joeandrews commented:
:tada: This PR is included in version 1.18.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

joeandrews commented:
:tada: This PR is included in version 1.20.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

joeandrews commented:
:tada: This PR is included in version 2.4.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

joeandrews commented:
:tada: This PR is included in version 1.6.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

joeandrews commented:
:tada: This PR is included in version 1.3.2 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

joeandrews commented:
:tada: This PR is included in version 1.1.4 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

joeandrews commented:
:tada: This PR is included in version 1.9.0 :tada:

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
